modules/programs/brew-bundle.nix

```diff
@@ -0,0 +1,213 @@
+        option will not be uninstalled automatically even if
+        <option>programs.brew-bundle.cleanupType</option> is set to <literal>"uninstall"</literal>
+        or <literal>"zap"</literal> (this is currently a limitation of Homebrew Bundle).
+
+        For more information on <command>mas</command> see: https://github.com/mas-cli/mas
+      '';
+    };
+
+    whalebrews = mkOption {
+      type = with types; listOf str;
+      default = [];
+      example = [ "whalebrew/wget" ];
+      description = ''
+        Docker images to install using <command>whalebrew</command>.
+
+        When this option is used, <literal>"whalebrew"</literal> is automatically added to
+        <option>programs.brew-bundle.brews</option>.
+
+        For more information on <command>whalebrew</command> see:
+        https://github.com/whalebrew/whalebrew
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        # 'brew tap' with custom Git URL
+        tap "user/tap-repo", "https://user@bitbucket.org/user/homebrew-tap-repo.git"
+
+        # set arguments for all 'brew cask install' commands
+        cask_args appdir: "~/Applications", require_sha: true
+
+        # 'brew install --with-rmtp', 'brew services restart' on version changes
+        brew "denji/nginx/nginx-full", args: ["with-rmtp"], restart_service: :changed
+        # 'brew install', always 'brew services restart', 'brew link', 'brew unlink mysql' (if it is installed)
+        brew "mysql@5.6", restart_service: true, link: true, conflicts_with: ["mysql"]
+
+        # 'brew cask install --appdir=~/my-apps/Applications'
+        cask "firefox", args: { appdir: "~/my-apps/Applications" }
+        # 'brew cask install' only if '/usr/libexec/java_home --failfast' fails
+        cask "java" unless system "/usr/libexec/java_home --failfast"
+      '';
+      description = "Extra lines to be added verbatim to the generated Brewfile.";
+    };
+  };
+
+  config = {
+    assertions = mkIf cfg.enable [
+      {
+        assertion = builtins.pathExists /usr/local/bin/brew;
+        message = ''
+          Homebrew not installed.
+
+          Please install Homebrew yourself before using the programs.brew-bundle module.
+
+          See installation instructions at: https://brew.sh
+        '';
+      }
+    ];
+
+    programs.brew-bundle.brews =
+      optional (cfg.masApps != {}) "mas" ++
+      optional (cfg.whalebrews != []) "whalebrew";
+
+    environment.variables = mkIf cfg.enable (
+      (if cfg.setNoLockEnvvar then { HOMEBREW_BUNDLE_NO_LOCK = "1"; } else {}) //
+      (if cfg.setBrewfileEnvvar then { HOMEBREW_BUNDLE_FILE = "${brewfile}"; } else {})
+    );
+
+    system.activationScripts.brew-bundle.text = mkIf cfg.enable ''
+      # Homebrew Bundle
+      echo >&2 "Homebrew bundle..."
+      PATH=/usr/local/bin:$PATH brew update > /dev/null
+      PATH=/usr/local/bin:$PATH brew bundle ${brew-bunble-options}
```

Comment:
Currently during activation `brew update` is run before `brew bundle`. I set it up that way since I wanted Homebrew to update things whenever I ran `darwin-rebuild switch`. This obviously isn't idempotent, so maybe the activation script shouldn't run `brew update`, and `brew bundle` should run with the `HOMEBREW_NO_AUTO_UPDATE` environment variable set. 

Does that sound right? If so, thoughts on adding an option to enable `brew update` on activation?

Reply:
Personally, the sluggishness of `brew update` is one of the things that drove me toward nix in the first place, so I'd like to see this be handled manually, or to at least be the default if that's an option. Similar to a flake/niv/nix-channel update, I don't want packages to change unless I specify it myself.

Reply:
Alrighty, I just pushed a commit that removes the invocation of `brew update`, and adds an option `noAutoUpdate` which is defaults to `true`. For the reason's you mentioned, as well as idempotence, this seems like the obvious choice. It may even be the case that it's not worth including the `noAutoUpdate` option at all and just hardcoding this behavior, but I'm inclined to leave it in for now and defer to whatever @LnL7 thinks is best when they have a chance to way in.

Reply:
Can I propose `autoUpdate = false` instead of `noAutoUpdate = true`?

Reply:
This is another example of me trying to stick to using the terms that Homebrew Bundle itself uses, where in this case it's the `HOMEBREW_NO_AUTO_UPDATE` environment variable. In this case though I'd be pretty happy to change it as you suggest if others agree.

Reply:
Thinking about it more, I've updated that the option should indeed be renamed to `autoUpdate` as @eraserhd suggested, because double negatives are the worst.